60201359 이주영 9주차 실습

In [7]:
!pip install mlxtend

In [21]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [22]:
from sklearn import datasets
df=pd.read_csv("data/contents.csv")
print(df.shape)
df.head()

(26, 3)


ID       일자   시청컨텐츠
0            jenny  07월 01일    겨울왕국
1            jisun  07월 02일  분노의 질주
2       peter_1021  07월 03일    스타워즈
3          young60  07월 05일    타이타닉
4  machine_learner  07월 06일  쥬라기 월드

In [23]:
df = df.drop(['일자'], axis=1)
df

ID   시청컨텐츠
0             jenny    겨울왕국
1             jisun  분노의 질주
2        peter_1021    스타워즈
3           young60    타이타닉
4   machine_learner  쥬라기 월드
5   machine_learner     아바타
6   machine_learner    해리포터
7             jisun    어벤져스
8            jlover    스타워즈
9        peter_1021    어벤져스
10            jenny  미녀와 야수
11           jlover     아바타
12            jenny  쥬라기 월드
13          eclipse    스타워즈
14          eclipse    어벤져스
15            jenny     아바타
16            jenny    타이타닉
17       peter_1021    아이언맨
18          eclipse  분노의 질주
19           jlover    라이온킹
20          young60    스타워즈
21          young60    겨울왕국
22           jlover    어벤져스
23          young60  미녀와 야수
24          eclipse  쥬라기 월드
25          young60    라이온킹

In [24]:
df[["ID","시청컨텐츠"]].values.tolist()

[['jenny', '겨울왕국'],
 ['jisun', '분노의 질주'],
 ['peter_1021', '스타워즈'],
 ['young60', '타이타닉'],
 ['machine_learner', '쥬라기 월드'],
 ['machine_learner', '아바타'],
 ['machine_learner', '해리포터'],
 ['jisun', '어벤져스'],
 ['jlover', '스타워즈'],
 ['peter_1021', '어벤져스'],
 ['jenny', '미녀와 야수'],
 ['jlover', '아바타'],
 ['jenny', '쥬라기 월드'],
 ['eclipse', '스타워즈'],
 ['eclipse', '어벤져스'],
 ['jenny', '아바타'],
 ['jenny', '타이타닉'],
 ['peter_1021', '아이언맨'],
 ['eclipse', '분노의 질주'],
 ['jlover', '라이온킹'],
 ['young60', '스타워즈'],
 ['young60', '겨울왕국'],
 ['jlover', '어벤져스'],
 ['young60', '미녀와 야수'],
 ['eclipse', '쥬라기 월드'],
 ['young60', '라이온킹']]

In [27]:
dataset = []
for name in df.ID.unique():
    name_mov = list(df[df['ID']==name]['시청컨텐츠'])
    dataset.append(name_mov)
dataset

[['겨울왕국', '미녀와 야수', '쥬라기 월드', '아바타', '타이타닉'],
 ['분노의 질주', '어벤져스'],
 ['스타워즈', '어벤져스', '아이언맨'],
 ['타이타닉', '스타워즈', '겨울왕국', '미녀와 야수', '라이온킹'],
 ['쥬라기 월드', '아바타', '해리포터'],
 ['스타워즈', '아바타', '라이온킹', '어벤져스'],
 ['스타워즈', '어벤져스', '분노의 질주', '쥬라기 월드']]

In [28]:
#TransactionEncoder는 리스트의 리스트로 구성된 트랜잭션 데이터를 Numpy 배열로 인코딩하기 위함
te=TransactionEncoder()
te_ary = te.fit(dataset).transform(dataset)
te_ary

array([[ True, False,  True, False, False,  True, False, False,  True,
         True, False],
       [False, False, False,  True, False, False, False,  True, False,
        False, False],
       [False, False, False, False,  True, False,  True,  True, False,
        False, False],
       [ True,  True,  True, False,  True, False, False, False, False,
         True, False],
       [False, False, False, False, False,  True, False, False,  True,
        False,  True],
       [False,  True, False, False,  True,  True, False,  True, False,
        False, False],
       [False, False, False,  True,  True, False, False,  True,  True,
        False, False]])

In [29]:
te.columns_

['겨울왕국',
 '라이온킹',
 '미녀와 야수',
 '분노의 질주',
 '스타워즈',
 '아바타',
 '아이언맨',
 '어벤져스',
 '쥬라기 월드',
 '타이타닉',
 '해리포터']

In [30]:
df = pd.DataFrame(te_ary, columns=te.columns_)
df

겨울왕국   라이온킹  미녀와 야수  분노의 질주   스타워즈    아바타   아이언맨   어벤져스  쥬라기 월드   타이타닉  \
0   True  False    True   False  False   True  False  False    True   True   
1  False  False   False    True  False  False  False   True   False  False   
2  False  False   False   False   True  False   True   True   False  False   
3   True   True    True   False   True  False  False  False   False   True   
4  False  False   False   False  False   True  False  False    True  False   
5  False   True   False   False   True   True  False   True   False  False   
6  False  False   False    True   True  False  False   True    True  False   

    해리포터  
0  False  
1  False  
2  False  
3  False  
4   True  
5  False  
6  False

In [60]:
#지지도 확인
frequent_itemsets= apriori(df,min_support = 0.2, use_colnames=True)
frequent_itemsets

support              itemsets
0   0.285714                (겨울왕국)
1   0.285714                (라이온킹)
2   0.285714              (미녀와 야수)
3   0.285714              (분노의 질주)
4   0.571429                (스타워즈)
5   0.428571                 (아바타)
6   0.571429                (어벤져스)
7   0.428571              (쥬라기 월드)
8   0.285714                (타이타닉)
9   0.285714        (겨울왕국, 미녀와 야수)
10  0.285714          (겨울왕국, 타이타닉)
11  0.285714          (스타워즈, 라이온킹)
12  0.285714        (타이타닉, 미녀와 야수)
13  0.285714        (어벤져스, 분노의 질주)
14  0.428571          (어벤져스, 스타워즈)
15  0.285714         (쥬라기 월드, 아바타)
16  0.285714  (겨울왕국, 타이타닉, 미녀와 야수)

In [61]:
#지지도 내림차순 정렬
frequent_itemsets.sort_values('support', ascending=False)

support              itemsets
4   0.571429                (스타워즈)
6   0.571429                (어벤져스)
14  0.428571          (어벤져스, 스타워즈)
5   0.428571                 (아바타)
7   0.428571              (쥬라기 월드)
0   0.285714                (겨울왕국)
11  0.285714          (스타워즈, 라이온킹)
15  0.285714         (쥬라기 월드, 아바타)
13  0.285714        (어벤져스, 분노의 질주)
12  0.285714        (타이타닉, 미녀와 야수)
8   0.285714                (타이타닉)
10  0.285714          (겨울왕국, 타이타닉)
9   0.285714        (겨울왕국, 미녀와 야수)
1   0.285714                (라이온킹)
3   0.285714              (분노의 질주)
2   0.285714              (미녀와 야수)
16  0.285714  (겨울왕국, 타이타닉, 미녀와 야수)

In [69]:
#신뢰도 확인
#설명 = shift + tap
association_rules(frequent_itemsets, metric ="confidence", min_threshold=1)

antecedents     consequents  antecedent support  consequent support  \
0           (겨울왕국)        (미녀와 야수)            0.285714            0.285714   
1         (미녀와 야수)          (겨울왕국)            0.285714            0.285714   
2           (겨울왕국)          (타이타닉)            0.285714            0.285714   
3           (타이타닉)          (겨울왕국)            0.285714            0.285714   
4           (라이온킹)          (스타워즈)            0.285714            0.571429   
5           (타이타닉)        (미녀와 야수)            0.285714            0.285714   
6         (미녀와 야수)          (타이타닉)            0.285714            0.285714   
7         (분노의 질주)          (어벤져스)            0.285714            0.571429   
8     (겨울왕국, 타이타닉)        (미녀와 야수)            0.285714            0.285714   
9   (겨울왕국, 미녀와 야수)          (타이타닉)            0.285714            0.285714   
10  (타이타닉, 미녀와 야수)          (겨울왕국)            0.285714            0.285714   
11          (겨울왕국)  (타이타닉, 미녀와 야수)            0.285714            0.285714   
12          (타이타닉)  (겨울왕국, 미녀와 야수)            0.285714            0.285714   
13        (미녀와 야수)    (겨울왕국, 타이타닉)            0.285714            0.285714   

     support  confidence  lift  leverage  conviction  zhangs_metric  
0   0.285714         1.0  3.50  0.204082         inf            1.0  
1   0.285714         1.0  3.50  0.204082         inf            1.0  
2   0.285714         1.0  3.50  0.204082         inf            1.0  
3   0.285714         1.0  3.50  0.204082         inf            1.0  
4   0.285714         1.0  1.75  0.122449         inf            0.6  
5   0.285714         1.0  3.50  0.204082         inf            1.0  
6   0.285714         1.0  3.50  0.204082         inf            1.0  
7   0.285714         1.0  1.75  0.122449         inf            0.6  
8   0.285714         1.0  3.50  0.204082         inf            1.0  
9   0.285714         1.0  3.50  0.204082         inf            1.0  
10  0.285714         1.0  3.50  0.204082         inf            1.0  
11  0.285714         1.0  3.50  0.204082         inf            1.0  
12  0.285714         1.0  3.50  0.204082         inf            1.0  
13  0.285714         1.0  3.50  0.204082         inf            1.0

In [70]:
#향상도 확인
association_rules(frequent_itemsets, metric="lift",min_threshold=1)

antecedents     consequents  antecedent support  consequent support  \
0           (겨울왕국)        (미녀와 야수)            0.285714            0.285714   
1         (미녀와 야수)          (겨울왕국)            0.285714            0.285714   
2           (겨울왕국)          (타이타닉)            0.285714            0.285714   
3           (타이타닉)          (겨울왕국)            0.285714            0.285714   
4           (스타워즈)          (라이온킹)            0.571429            0.285714   
5           (라이온킹)          (스타워즈)            0.285714            0.571429   
6           (타이타닉)        (미녀와 야수)            0.285714            0.285714   
7         (미녀와 야수)          (타이타닉)            0.285714            0.285714   
8           (어벤져스)        (분노의 질주)            0.571429            0.285714   
9         (분노의 질주)          (어벤져스)            0.285714            0.571429   
10          (어벤져스)          (스타워즈)            0.571429            0.571429   
11          (스타워즈)          (어벤져스)            0.571429            0.571429   
12        (쥬라기 월드)           (아바타)            0.428571            0.428571   
13           (아바타)        (쥬라기 월드)            0.428571            0.428571   
14    (겨울왕국, 타이타닉)        (미녀와 야수)            0.285714            0.285714   
15  (겨울왕국, 미녀와 야수)          (타이타닉)            0.285714            0.285714   
16  (타이타닉, 미녀와 야수)          (겨울왕국)            0.285714            0.285714   
17          (겨울왕국)  (타이타닉, 미녀와 야수)            0.285714            0.285714   
18          (타이타닉)  (겨울왕국, 미녀와 야수)            0.285714            0.285714   
19        (미녀와 야수)    (겨울왕국, 타이타닉)            0.285714            0.285714   

     support  confidence      lift  leverage  conviction  zhangs_metric  
0   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
1   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
2   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
3   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
4   0.285714    0.500000  1.750000  0.122449    1.428571       1.000000  
5   0.285714    1.000000  1.750000  0.122449         inf       0.600000  
6   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
7   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
8   0.285714    0.500000  1.750000  0.122449    1.428571       1.000000  
9   0.285714    1.000000  1.750000  0.122449         inf       0.600000  
10  0.428571    0.750000  1.312500  0.102041    1.714286       0.555556  
11  0.428571    0.750000  1.312500  0.102041    1.714286       0.555556  
12  0.285714    0.666667  1.555556  0.102041    1.714286       0.625000  
13  0.285714    0.666667  1.555556  0.102041    1.714286       0.625000  
14  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
15  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
16  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
17  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
18  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
19  0.285714    1.000000  3.500000  0.204082         inf       1.000000

In [71]:
df_assoc = association_rules(frequent_itemsets, metric="lift", min_threshold=0.3)
df_assoc.sort_values('lift', ascending =False) #lift 기준 내림차순 정렬

antecedents     consequents  antecedent support  consequent support  \
0           (겨울왕국)        (미녀와 야수)            0.285714            0.285714   
1         (미녀와 야수)          (겨울왕국)            0.285714            0.285714   
18          (타이타닉)  (겨울왕국, 미녀와 야수)            0.285714            0.285714   
17          (겨울왕국)  (타이타닉, 미녀와 야수)            0.285714            0.285714   
16  (타이타닉, 미녀와 야수)          (겨울왕국)            0.285714            0.285714   
15  (겨울왕국, 미녀와 야수)          (타이타닉)            0.285714            0.285714   
14    (겨울왕국, 타이타닉)        (미녀와 야수)            0.285714            0.285714   
19        (미녀와 야수)    (겨울왕국, 타이타닉)            0.285714            0.285714   
7         (미녀와 야수)          (타이타닉)            0.285714            0.285714   
6           (타이타닉)        (미녀와 야수)            0.285714            0.285714   
3           (타이타닉)          (겨울왕국)            0.285714            0.285714   
2           (겨울왕국)          (타이타닉)            0.285714            0.285714   
8           (어벤져스)        (분노의 질주)            0.571429            0.285714   
9         (분노의 질주)          (어벤져스)            0.285714            0.571429   
5           (라이온킹)          (스타워즈)            0.285714            0.571429   
4           (스타워즈)          (라이온킹)            0.571429            0.285714   
12        (쥬라기 월드)           (아바타)            0.428571            0.428571   
13           (아바타)        (쥬라기 월드)            0.428571            0.428571   
11          (스타워즈)          (어벤져스)            0.571429            0.571429   
10          (어벤져스)          (스타워즈)            0.571429            0.571429   

     support  confidence      lift  leverage  conviction  zhangs_metric  
0   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
1   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
18  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
17  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
16  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
15  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
14  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
19  0.285714    1.000000  3.500000  0.204082         inf       1.000000  
7   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
6   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
3   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
2   0.285714    1.000000  3.500000  0.204082         inf       1.000000  
8   0.285714    0.500000  1.750000  0.122449    1.428571       1.000000  
9   0.285714    1.000000  1.750000  0.122449         inf       0.600000  
5   0.285714    1.000000  1.750000  0.122449         inf       0.600000  
4   0.285714    0.500000  1.750000  0.122449    1.428571       1.000000  
12  0.285714    0.666667  1.555556  0.102041    1.714286       0.625000  
13  0.285714    0.666667  1.555556  0.102041    1.714286       0.625000  
11  0.428571    0.750000  1.312500  0.102041    1.714286       0.555556  
10  0.428571    0.750000  1.312500  0.102041    1.714286       0.555556

지지도가 가장 높은 조합은 스타워즈 -> 어벤져스이다.(지지도 상위 42%)
신뢰도는 75%, 향상도는 1.32 = 1이상으로 추천할 만 함
전체 기록 중에 스타워즈와 어벤져스를 모두 시청한 비율은 42%이며 스타워즈를 본 사람 중 75%는 이후 어벤져스를 시청하고,
이를 이용하여 마케팅을 하면 이 연관분석 결과를 모를 때보다 효과가 32% 증가한다는 것을 의미한다.

향상도를 기준으로 세가지의 조합이 나온 것 중 타이타닉, 겨울왕국, 미녀와 야수의 향상도도 1 이상으로 추천할 만한 조합이다. 위 조합을 적용하여 마케팅을 하면 시청효과가 3.5배 상승한다.